File: 06_strings/06_10_roman_to_integer/roman_to_integer.cpp
#include <iostream>
#include <unordered_map>

int romanToInt(const std::string& s) {
    if (s.empty()) {
        return -1;
    }

    std::unordered_map<char, int> romanNumerals = {
        {'I', 1},
        {'V', 5},
        {'X', 10},
        {'L', 50},
        {'C', 100},
        {'D', 500},
        {'M', 1000}
    };

    int result = 0;
    for (size_t i = 0; i < s.length() - 1; i++) {
        int current = romanNumerals[s[i]];
        int next = romanNumerals[s[i + 1]];
        if (current < next) {
            result -= current;
        } else {
            result += current;
        }
    }
    result += romanNumerals[s.back()];

    return result;
}

int main() {
    // Test cases
    std::cout << "Input: III -> Output: " << romanToInt("III") << std::endl;
    std::cout << "Input: IV -> Output: " << romanToInt("IV") << std::endl;
    std::cout << "Input: IX -> Output: " << romanToInt("IX") << std::endl;
    std::cout << "Input: LVIII -> Output: " << romanToInt("LVIII") << std::endl;
    std::cout << "Input: MCMXCIV -> Output: " << romanToInt("MCMXCIV") << std::endl;

    return 0;
}

File: 06_strings/06_10_roman_to_integer/roman_to_integer.py
def roman_to_int(s):
    if not s:
        return -1

    roman_numerals = {
        'I': 1,
        'V': 5,
        'X': 10,
        'L': 50,
        'C': 100,
        'D': 500,
        'M': 1000
    }
    
    result = 0
    
    for i in range(len(s) - 1):
        current_value = roman_numerals[s[i]]
        next_value = roman_numerals[s[i + 1]]
        if current_value < next_value:
            result -= current_value
        else:
            result += current_value
    result += roman_numerals[s[(len(s) - 1)]]
    return result

# Test cases
print(f"Input: III -> Output: {roman_to_int('III')}")
print(f"Input: IV -> Output: {roman_to_int('IV')}")
print(f"Input: IX -> Output: {roman_to_int('IX')}")
print(f"Input: LVIII -> Output: {roman_to_int('LVIII')}")
print(f"Input: MCMXCIV -> Output: {roman_to_int('MCMXCIV')}")